### PR TITLE
Preserve document ordering of entities

### PR DESCRIPTION
**Pull request by justinfagnani**

### CODE REVIEW

File: src/javascript/javascript-document.ts [hunk 1]
```diff
@@ -45,4 +45,14 @@ export class JavaScriptDocument extends Document<Program, Visitor> {
       fallback: 'iteration',
     });
   }
+
+  forEachNode(callback: (node: Node) => void) {
+    traverse(<Node><any>this.ast, {
```
Comment (rictic) [inline src/javascript/javascript-document.ts]: Why do you need to do the `<Node><any>` dance, a Program is a Node..

File: src/entity/find-entities.ts [hunk 2]
```diff
@@ -86,5 +86,37 @@ export async function findEntities(
   }
 
   // Flatten the nested list
-  return Array.prototype.concat.apply([], nestedEntities);
+  let unorderedEntities = Array.prototype.concat.apply([], nestedEntities);
+  let orderedEntities = orderEntities(document, unorderedEntities);
```
Comment (rictic) [inline src/entity/find-entities.ts]: WDYT about leaving out the `orderedEntities` local and just returning instead?

File: src/css/css-document.ts [hunk 3]
```diff
@@ -22,4 +22,8 @@ export class CssDocument extends Document<Node, NodeVisitor> {
   visit(visitors: NodeVisitor[]) {
     throw new Error('Not implemented');
   }
+
+  forEachNode(callback: (node: Node) => void) {
```
Comment (rictic) [inline src/css/css-document.ts]:
```javascript
/**
 * _Must_ call the callback with every node, and must do so in document order.
 */
```

File: src/entity/find-entities.ts [hunk 4]
```diff
@@ -86,5 +86,37 @@ export async function findEntities(
...
+
+  return orderedEntities;
+}
+
+function orderEntities(
+  document: Document<any, any>,
+  unorderedEntities: Descriptor[]): Descriptor[] {
+
+  let entitiesByNode = new Map<any, Descriptor[]>();
+  let orderedEntities: Descriptor[] = [];
```
Comment (rictic) [inline src/entity/find-entities.ts]: This type should be `Descriptor[][]` shouldn't it? Is there something wrong with the type checker, or something wrong with my brain?

**Comment by rictic** [comment]
Looks reasonable to me. You mentioned adding more tests, and that seems wise to me too.

File: src/javascript/javascript-document.ts (hunk 1 above)
Comment (justinfagnani) [inline src/javascript/javascript-document.ts]: I forget, let me check

Comment (justinfagnani) [inline src/javascript/javascript-document.ts]: removed

File: src/entity/find-entities.ts (hunk 2 above)
Comment (justinfagnani) [inline src/entity/find-entities.ts]: sure

File: src/css/css-document.ts (hunk 3 above)
Comment (justinfagnani) [inline src/css/css-document.ts]: added to comment

File: src/entity/find-entities.ts (hunk 4 above)
Comment (justinfagnani) [inline src/entity/find-entities.ts]: You're right. I don't know what's going on with the type checking here. I've tried several combinations of what should be invalid types and tsc isn't complaining.

**Comment by justinfagnani** [comment]
I want to move on this before adding more test cases so we can unlock bundling...

**Pull request merged**